fairlearn/postprocessing/_cvxpy_threshold_optimizer.py

```diff
@@ -0,0 +1,395 @@
+"""Threshold optimizer with relaxed fairness constraints.
+
+TODO
+----
+- Add option for constraining only equality of FPR or TPR (currently it must be 
+both -> equal odds);
+- Add option for constraining equality of positive predictions (independence
+criterion, aka demographic parity);
+- Add option to use l1 or linf distances for maximum tolerance between points.
+  - Currently 'equal_odds' is defined using l-infinity distance (max between
+  TPR and FPR distances);
+
+"""
+import logging
+from itertools import product
+from typing import Callable
+
+import numpy as np
+from sklearn.metrics import roc_curve
+from sklearn.base import BaseEstimator, MetaEstimatorMixin
+
+from fairlearn.utils._input_validation import _validate_and_reformat_input
+from fairlearn.reductions._moments.error_rate import _MESSAGE_BAD_COSTS
+
+from ._cvxpy_utils import compute_equal_odds_optimum
+from ._roc_utils import (
+    roc_convex_hull,
+    calc_cost_of_point,
+)
+from ._randomized_classifiers import (  # TODO: try to use fairlearn's InterpolatedThreshold instead of our classifier API
+    RandomizedClassifier,
+    EnsembleGroupwiseClassifiers,
+)
+
+
+class _RelaxedThresholdOptimizer(BaseEstimator, MetaEstimatorMixin):
+    r"""Class to encapsulate all the logic needed to compute the optimal
+    postprocessing to fulfill fairness constraints with some optional
+    tolerance.
+
+    The method amounts to finding the set of (potentially randomized) 
+    group-specific decision thresholds that maximize some goal (e.g., accuracy),
+    given a maximum tolerance (or slack) on the fairness constraint fulfillment.
+    
+    This optimization problem amounts to a Linear Program (LP) as detailed in 
+    [1]_. Solving the LP requires installing `cvxpy`.
+
+    Read more in the :ref:`User Guide <postprocessing>`.
+
+    Parameters
+    ----------
+    estimator : object
+        A `scikit-learn compatible estimator <https://scikit-learn.org/stable/developers/develop.html#estimators>`_  # noqa
+        whose output will be postprocessed.
+        The estimator should output real-valued scores, as postprocessing 
+        results will be extremely poor when performed over binarized 
+        predictions.
+
+    tolerance : float
+        The absolute tolerance for the equal odds fairness constraint.
+        Will allow for `tolerance` difference between group-wise ROC points.
+
+    objective_costs : dict
+        A dictionary detailing the cost for false positives and false negatives,
+        of the form :code:`{'fp': <fp_cost>, 'fn': <fn_cost>}`. Will use the 0-1
+        loss by default (maximum accuracy).
+
+    grid_size : int, optional
```

Comment:
Right, I think this implementation is different in the sense that we only define the edges of the ROC curve, and any point in the interior of its convex hull is attainable.
Because the problem is defined as an LP, the domain is (allowed to be) a continuous space instead of a discrete grid.

Should I switch this back to `max_roc_ticks` ? It is still a grid size, just not the same grid as in TO.

Reply:
I suspect you're right in naming it something else. That would confuse people if both were `grid_size`, as it already confused me 😆 
I guess I should re-read the paper, but even in TO it's continuous. Our implementation is just a bit lazy and we take the closest grid point rather than finding the exact solution.

Reply:
The difference is a bit nuanced. AFAIK the `grid_size` in fairlearn's TO is the number of linear combinations between thresholds. In the RelaxedTO it is the number of thresholds (per group ROC).

1. We use these thresholds (and their associated ROC points) to create a multi-dimensional envelope (two dimensions per group; while in strict equalized odds it is only 2 dimensions in total because all ROC hulls are intersected).
2. Each group-specific "envelope" (ROC hull) is passed to cvxpy in the form of linear inequalities, hence all points in the interior of these ROC hulls are possible (there is no discretization at this point, the discretization happened before when figuring out the boundaries of the ROC hull, as the boundary polygon will have at most `grid_size` points).
3. cvxpy computes a solution in ROC space (ROC points for each group), and _then_ we try to bring that solution back to the "realizable" space of thresholds.

So the optimization is really performed in a continuous space, the `grid_size` is the maximum number of points in the envelope that constrains each group's optimization domain but any linear combination within these polygons is possible.